src/gcp_bigquery.rs

```diff
@@ -0,0 +1,449 @@
+use crate::{
+    csv::CsvWriter,
+    datasource::DatasourceWritable,
+    types::{IndexerContractMapping, IndexerGcpBigQueryConfig},
+};
+use async_trait::async_trait;
+use csv::ReaderBuilder;
+use gcp_bigquery_client::{
+    error::BQError,
+    model::{
+        table::Table, table_data_insert_all_request::TableDataInsertAllRequest,
+        table_field_schema::TableFieldSchema, table_schema::TableSchema,
+    },
+    Client,
+};
+use indexmap::IndexMap;
+use phf::phf_ordered_map;
+use polars::prelude::*;
+use serde_json::Value;
+use std::{any::Any, collections::HashMap, fs::File};
+
+///
+/// Columns common to all tables
+/// Each table's columns will include the following commmon fields
+///
+static COMMON_COLUMNS: phf::OrderedMap<&'static str, &'static str> = phf_ordered_map! {
+    "indexed_id" => "string",  // will need to generate uuid in rust; postgres allows for autogenerate
+    "contract_address" => "string",
+    "tx_hash" => "string",
+    "block_number" => "int",
+    "block_hash" => "string",
+    "timestamp" => "int"
+};
+
+/// Query client
+/// Impl for this struct is further below
+pub struct GcpBigQueryClient {
+    client: Client,
+    project_id: String,
+    dataset_id: String,
+    drop_tables: bool,
+    table_map: HashMap<String, IndexMap<String, String>>,
+}
+
+#[derive(Debug)]
+pub enum GcpClientError {
+    MissingParameters,
+    Other(String),
+    BigQueryError(BQError),
+}
+
+///
+/// Factory function to initialize the GCP bigquery client
+/// Also deletes / creates tables as necessary (controlled via setting)
+///
+/// # Arguments
+///
+/// * `indexer_gcp_bigquery_config` - gcp config block
+/// * `event_mappings` - list of event mapping blocks (all events to parse)
+pub async fn init_gcp_bigquery_db(
+    indexer_gcp_bigquery_config: &IndexerGcpBigQueryConfig,
+    event_mappings: &[IndexerContractMapping],
+) -> Result<GcpBigQueryClient, GcpClientError> {
+    let gcp_bigquery = GcpBigQueryClient::new(&indexer_gcp_bigquery_config, &event_mappings)
+        .await
+        .unwrap();
```

Comment:
can you not use ? here else it would panic - if error is not a GcpClientError you could parse it into that?

Reply:
missed this comment?

Reply:
done - yes, appears that some comments got minimized, thanks